enterprise/ha/src/main/java/org/neo4j/kernel/ha/com/master/ConversationManager.java

```diff
@@ -110,12 +127,22 @@ public void end( RequestContext context )
 
     public Set<RequestContext> getActiveContexts()
     {
-        return conversations.keys();
+        return conversations != null ? conversations.keys() : Collections.<RequestContext>emptySet() ;
     }
 
-    public void remove( RequestContext context )
+    /**
+     * Stop conversation for specified context.
+     * Conversation will still hold all already acquired locks, but will release all waiters and it will be
+     * impossible to get new locks out of it.
+     * @param context - context for which conversation should be stopped
+     */
+    public void stop( RequestContext context )
     {
-        conversations.remove( context );
+        Conversation conversation = conversations.end( context );
+        if ( conversation != null && conversation.isActive() )
+        {
+            conversation.stop();
+        }
```

Comment:
This is not good, Señor. When `end` returns, the value will be picked up asynchronously and fed to the reaper. The reaper will then close the conversation, which in turn closes the lock client, which then returns it to the pool of lock clients, wherefrom it can be claimed and used in another transaction. Thus, you might end up accidentally closing the lock client of a different conversation.